[DOCS] OAuth authentication added to SN connectors (#2048)
elastic/security-docs: docs/cases/cases-ui-integrations.asciidoc

* {ibm-r}
* {swimlane}

IMPORTANT: To create connectors and send cases to external systems, you need the

> https://www.elastic.co/subscriptions[appropriate license], and your role needs *All* privileges for the *Action and Connectors* feature. For more information, see <<case-permissions>>.

Reviewer: ```suggestion
https://www.elastic.co/subscriptions[appropriate license], and your role needs *All* privileges for the *Action and Connectors* feature. For more information, refer to <<case-permissions>>.
```

To push cases, you need to create a connector, which stores the information
required to interact with an external system.

After you have created a connector, you can set {es-sec} cases to
automatically close when they are sent to external systems.

[float]
[[create-new-connector]]
=== Create a new connector

. Go to *Investigate* -> *Cases* -> *Edit external connection*.
+
[role="screenshot"]
image::images/cases-ui-connector.png[Shows the page for creating connectors]
. From the *Incident management system* list, select *Add new connector*.

> . Select the system to send cases to: *{sn}*, *{jira}*, *{ibm-r}*, or *{swimlane}*.

Reviewer 2: There is setup that's required on the Servicenow instance before it can be used. Should those be included here? It's in the stack connector docs: https://www.elastic.co/guide/en/kibana/master/servicenow-action-type.html

Author: Are you referring to users needing to install the [Elastic for ITSM](https://store.servicenow.com/sn_appstore_store.do#!/store/application/7148dbc91bf1f450ced060a7234bcb88) and the [Elastic for Security Operations (SecOps)](https://store.servicenow.com/sn_appstore_store.do#!/store/application/2f0746801baeb01019ae54e4604bcb0f) applications from the ServiceNow Store before they can use the connectors?

Reviewer 2: No, for OAuth, they also need to [create an RSA keypair](https://www.elastic.co/guide/en/kibana/master/servicenow-action-type.html#_create_an_rsa_keypair_and_add_an_x_509_certificate) and [create an OAuth JWT API endpoint](https://www.elastic.co/guide/en/kibana/master/servicenow-action-type.html#_create_an_oauth_jwt_api_endpoint_for_external_clients_with_a_jwt_verifiers_map)

Author: Ah, gotcha. So I loosely refer to those sections in the description for the Use OAuth authentication toggle, but I wasn't able to link to either sections directly because they don't have anchor text at the top. I'll need to open a PR against those 3 topics in the Kibana repo to add the anchor text. 
cc: @lcawl

+
IMPORTANT: If you've upgraded from {stack} version 7.15.0 or earlier to version 7.16.0 or later, you must install https://store.servicenow.com/sn_appstore_store.do#!/store/application/7148dbc91bf1f450ced060a7234bcb88[Elastic for ITSM] or https://store.servicenow.com/sn_appstore_store.do#!/store/application/2f0746801baeb01019ae54e4604bcb0f[Elastic for Security Operations (SecOps)] on your {sn} instance and complete additional configuration steps before creating a new {sn} ITSM or SecOps connector. For more information, refer to {kibana-ref}/servicenow-sir-action-type.html[{sn} SecOps] and {kibana-ref}/servicenow-action-type.html[{sn} ITSM].

> . Enter your required settings.

Reviewer 3: This table contains _a lot_ of disparate info -- I'd imagine a user would only want information about the specific connector type they're working with, and could get frustrated having to weed out info about the other types.

Seems like a "[tabbed widget](https://github.com/elastic/docs#tabbed-widgets)" could be a good format for this, or at least just breaking into smaller, more focused tables for each connector type.

+
|===

| *Connector name* | Name for the connector.

| *URL* | The URL of the external system to which you want to send cases.

> | *{sn} instance URL* | ({sn} only) The URL of the {sn} instance to which you want to send cases.

Author: Hm, I'm not sure I understand your question about creating a separate entry for ServiceNow. To your point, I do think having all of these fields listed in step 4 can be overwhelming, as a lot of the information is specific to certain connectors. That said, I think the original author tried to convey as best as possible that by specifying in parentheses which connector the field was present on. So in this example, the docs indicate that the SN instance **URL** field is available on SN connectors only and the **URL** field isn't present on SN connectors. The **URL** field is available on IBM and Jira connectors though, so I can add that info in parentheses to further clarify if that helps.

> | *{sn} instance URL* | ({sn} only) The URL of the {sn} instance to which you want to send cases.

Reviewer 2: Yeah, that would be helpful. Right now it looks like *URL* is required for all connectors, including SN.

> | *{sn} instance URL* | ({sn} only) The URL of the {sn} instance to which you want to send cases.

Author: I'll make that change, thanks!

| *Use OAuth authentication* | ({sn} only) Enable the toggle to authenticate a connection between Elastic and {sn}. Refer to the {sn} {kibana-ref}/action-types.html[Connector] topics for more information.

| *API URL* |  ({swimlane} only) The URL of the {swimlane} instance to which you want to send cases.

| *Organization ID* | ({ibm-r} only) Your organization’s {ibm-r} ID number.

| *Application ID* | ({swimlane} only) The application ID of your {swimlane} application. From {swimlane}, you can find the application
ID by checking your application’s settings or at the end of your application’s URL after you’ve opened it.

| *Username* | ({sn} only) The username of the {sn} account used to access the {sn} instance.

| *Password* | ({sn} only) The password of the {sn} account is used to access the {sn} instance.

> | *Project key* | ({jira} only) The key of the {jira} project to which you are sending cases.

Reviewer 2: There are a bunch of fields that are required for ServiceNow if Use OAuth is true: Client ID, Client Secret, JWT Key ID, Private Key, Private Key Password. Should these all be added?

Author: Yep! I'll refer to the ServiceNow stack management docs for field definitions.

| *Email address* | ({jira} only) The {jira} account username or email.

| *API token* | ({jira} only) The API token or password is used to authenticate {jira} updates.

| *API key ID* | ({ibm-r} only) The API key is used to authenticate {ibm-r} updates.

| *API key secret* | ({ibm-r} only) The API key secret is used to authenticate {ibm-r} updates.

| *API token* | ({swimlane} only) The {swimlane} API authentication token is used for HTTP Basic authentication.
This is the personal access token for your user role.

|===
+
. Choose the connector type ({swimlane} only):
+
|===

| *All* | You can choose to set all or no field mappings when creating your new {swimlane} connector. However, note that if
you don’t set field mappings now, you’ll be prompted to do so if you want to use the connector for a case or a rule. This

> is because the prompts no longer display once you set up the required mappings.

Author: Agree, the last sentence can be removed. It's typical behavior for prompts to stop once requirements are fulfilled.

| *Alerts* | Provide an alert ID and rule name.

| *Cases* | Provide a case ID, a case name, comments, and a description.

|===
+
. Save the connector.

TIP: To see how to connect {elastic-sec} to {jira}, watch the <<connect-security-to-jira, tutorial>> at the end of this topic.

[float]
[[mapped-case-fields]]
=== Mapped case fields

To represent an {es-sec} case in an external system, {es-sec} case fields are
mapped as follows:

NOTE: Data from mapped case fields can be pushed to external systems but cannot be pulled in.

* For {sn} incidents:
+
|===

| *Title* | Mapped to the {sn} `Short description` field. When an update to a case title is sent to {sn}, the existing {sn} `Short description` field is overwritten.

| *Description* | Mapped to the {sn} `Description` field. When an update to a case description is sent to {sn}, the existing {sn} `Description` field is overwritten.

> | *Comments* | Mapped to the {sn} `Work Notes` field. When a comment is updated in a case, a new comment is added to the {sn} incident.

Reviewer 3: There's a lot of repetition in this and the next three tables, so the more important info (A is mapped to B) gets lost in a lot of text. 
 
Would it be possible to add a column for "Mapped field" for all of these tables? That would relieve some of the repetition of "Mapped to the XXX `yyy` field" in every line, and it'd be instantly clear what is mapped to what. I know our narrow page margins make tables difficult, but maybe just one more column could still work...?

|===
+

* For {jira} issues:
+
|===

| *Title* | Mapped to the {jira} `Summary` field. When an update to a case title is sent to {jira}, the existing {jira} `Summary` field is overwritten.

| *Description* | Mapped to the {jira} `Description` field. When an update to a case description is sent to {jira}, the existing {jira} `Description` field is overwritten.

| *Comments* | Mapped to the {jira} `Comments` field. When a comment is updated in a case, a new comment is added to the {jira} incident.

|===
+

* For {ibm-r} issues:
+
|===

> | *Title* | Mapped to the {ibm-r} `Name` field. When an update to a case title is sent to {ibm-r}, the existing {ibm-r} `Name` field is overwritten.

Author: @benironside are suggesting something along the lines of:

| *Title* | Mapped to and overwrites the {ibm-r} `Name` field.

[float]
[[create-new-connector]]
=== Create a new connector

. Go to *Investigate* -> *Cases* -> *Edit external connection*.
+
[role="screenshot"]
image::images/cases-ui-connector.png[Shows the page for creating connectors]
. From the *Incident management system* list, select *Add new connector*.
. Select the system to send cases to: *{sn}*, *{jira}*, *{ibm-r}*, or *{swimlane}*.

+
IMPORTANT: If you've upgraded from {stack} version 7.15.0 or earlier to version 7.16.0 or later, you must install https://store.servicenow.com/sn_appstore_store.do#!/store/application/7148dbc91bf1f450ced060a7234bcb88[Elastic for ITSM] or https://store.servicenow.com/sn_appstore_store.do#!/store/application/2f0746801baeb01019ae54e4604bcb0f[Elastic for Security Operations (SecOps)] on your {sn} instance and complete additional configuration steps before creating a new {sn} ITSM or SecOps connector. For more information, refer to {kibana-ref}/servicenow-action-type.html[{sn} ITSM] and {kibana-ref}/servicenow-sir-action-type.html[{sn} SecOps].

. Enter your required settings.
+
|===

> | *Connector name* | Name for the connector.

Reviewer 4: Just adding an article here for consistency. `The` vs `a` also works. 
```suggestion
| *Connector name* | A name for the connector.
```

| *URL* | ({ibm-r} and {jira} only) The URL of the external system to which you want to send cases.

| *{sn} instance URL* | ({sn} only) The URL of the {sn} instance to which you want to send cases.

> | *Use OAuth authentication* | ({sn} only) Enable the toggle to authenticate a connection between Elastic and {sn}. Refer to the {sn} {kibana-ref}/action-types.html[Connector] topics for more information.

Reviewer 4: ```suggestion
| *Use OAuth authentication* | ({sn} only) Turn on the toggle to authenticate a connection between Elastic and {sn}. Refer to the {sn} {kibana-ref}/action-types.html[Connector] topics for more information.
```

| *API URL* |  ({swimlane} only) The URL of the {swimlane} instance to which you want to send cases.

| *Organization ID* | ({ibm-r} only) Your organization’s {ibm-r} ID number.

| *Application ID* | ({swimlane} only) The application ID of your {swimlane} application. From {swimlane}, you can find the application
ID by checking your application’s settings or at the end of your application’s URL after you’ve opened it.

| *Username* | ({sn} only and displays if *Use OAuth authentication* is turned off) The username of the {sn} account used to access the {sn} instance.

| *Password* | ({sn} only and displays if *Use OAuth authentication* is turned off) The password of the {sn} account used to access the {sn} instance.

| *Client ID* | ({sn} only and displays if *Use OAuth authentication* is toggled on) The client ID assigned to your OAuth application.

> | *User Identifier* | ({sn} only and displays if *Use OAuth authentication* is toggled on) Identifier to use for OAuth type authentication. Use the value you entered into the *User field* when you created an OAuth JWT API endpoint for external clients.

Reviewer 4: ```suggestion
| *User Identifier* | ({sn} only and displays if *Use OAuth authentication* is turned on) Identifier to use for OAuth type authentication. Use the value you entered into the *User field* when you created an OAuth JWT API endpoint for external clients.
```

> | *JWT Verifier Key ID* | ({sn} only and displays if *Use OAuth authentication* is toggled on) The key ID assigned to the JWT Verifier Map of your OAuth application.

Reviewer 4: ```suggestion
| *JWT Verifier Key ID* | ({sn} only and displays if *Use OAuth authentication* is turned on) The key ID assigned to the JWT Verifier Map of your OAuth application.
```

> | *Client Secret* | ({sn} only and displays if *Use OAuth authentication* is toggled on) The client secret assigned to your OAuth application.

Reviewer 4: ```suggestion
| *Client Secret* | ({sn} only and displays if *Use OAuth authentication* is turned on) The client secret assigned to your OAuth application.
```

> | *Private Key* | ({sn} only and displays if *Use OAuth authentication* is toggled on) The RSA private key generated when you created an RSA keypair.

Reviewer 4: ```suggestion
| *Private Key* | ({sn} only and displays if *Use OAuth authentication* is turned on) The RSA private key generated when you created an RSA keypair.
```

> | *Private Key Password* | ({sn} only and displays only if *Use OAuth authentication* is toggled on) The The password for the RSA private key generated during setup, if set.

Reviewer 4: ```suggestion
| *Private Key Password* | ({sn} only and displays only if *Use OAuth authentication* is turned on) The The password for the RSA private key generated during setup, if set.
```
